File: src/etch/bindings/csharp/transport/fmt/binary/BinaryTaggedDataOutput.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Diagnostics;
using Etch.Msg;
using Etch.Util;

namespace Etch.Transport.Fmt.Binary
{
    /// <summary>
    /// Description of TaggedDataOutputStream.
    /// </summary>
    public sealed class BinaryTaggedDataOutput : BinaryTaggedData, TaggedDataOutput
    {
        /// <summary>
        /// Constructs the BinaryTaggedDataOutput.
        /// </summary>

        /// <param name="vf"></param>
        public BinaryTaggedDataOutput(ValueFactory vf)
            : base(vf)
        {

        }

        /// <summary>
        /// Constructs the BinaryTaggedDataOutput with a null buffer.
        /// </summary>
        /// <param name="vf">the value factory for the service.</param>
        /// <param name="uri">the uri used to construct the transport stack.</param>
        public BinaryTaggedDataOutput(ValueFactory vf, string uri)
            : base(vf)
        { }

        private FlexBuffer buf;

        //////////////////////////////
        // TaggedDataOutput methods //
        //////////////////////////////

        #region TaggedDataOutput methods

        public void SetBuffer(FlexBuffer buf)
        {
            this.buf = buf;
        }

        public void WriteMessage(Message msg, FlexBuffer buf)
        {
            try
            {
                SetBuffer(buf);
                StartMessage(msg);
                WriteKeysAndValues(msg);
                EndMessage(msg);
            }
            finally
            {
                SetBuffer(null);
            }
        }

        public void WriteStruct(StructValue sv)
        {
            StartStruct(sv);
            WriteKeysAndValues(sv);
            EndStruct(sv);
        }

        public void WriteArray(ArrayValue av, Validator v)
        {
            StartArray(av);
            WriteValues(av, v);
            EndArray(av);
        }

        private void WriteKeysAndValues(StructValue sv)
        {
            XType t = sv.GetXType;
            foreach (KeyValuePair<Field, Object> me in sv)
            {
                Field f = me.Key;
                WriteIntValue(f.Id);
                WriteValue(t.GetValidator(f), me.Value);
            }
        }

        private void WriteValues(ArrayValue av, Validator v)
        {
            Validator ev = v.ElementValidator();
            foreach (Object value in av)
                WriteValue(ev, value);
        }


        /////////////////////////
        // Main output methods //
        /////////////////////////

        /// <summary>
        /// 
        /// </summary>
        /// <param name="msg"></param>
        /// Exception:
        ///     throws Exception
        public void StartMessage(Message msg)
        {
            // since message is top-level, no type code is written to
            // indicate a message is starting. we do write a version number
            // to indicate this version of the binary tagged data output
            buf.PutByte(VERSION);
            StartStruct(msg);
            //       WriteType( msg.GetXType );
        }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="msg"></param>
        /// Exception:
        ///     throws Exception
        public void EndMessage(Message msg)
        {
            //WriteValue( noneValidator, NONE );
            EndStruct(msg);

        }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="sv"></param>
        /// Exception:
        ///     throws Exception
        public void StartStruct(StructValue sv)
        {
            // the caller has already written a type code to indicate a
            // struct is starting.
            WriteType(sv.GetXType);
            WriteIntValue(sv.Count);
        }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="sv"></param>
        /// Exception:
        ///     throws Exception
        public void EndStruct(StructValue sv)
        {
            //    WriteValue( noneValidator, NONE );
            WriteNoneValue();
        }

        private readonly IList<StructValue> structs = new List<StructValue>();

        public void StartArray(ArrayValue array)
        {
            // the caller has already written a type code to indicate an
            // array is starting.
            sbyte type = array.Type;
            buf.PutByte(type);
            if (type == TypeCode.CUSTOM)
                WriteType(array.CustomStructType);
            //        WriteValue( intValidator, array.Dim );
            WriteIntValue(array.Dim);
            WriteIntValue(array.Size());

        }

        private void WriteType(XType type)
        {
            WriteIntValue(type.Id);
        }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="array"></param>
        /// Exception:
        ///     throws Exception
        public void EndArray(ArrayValue array)
        {
            WriteNoneValue();
        }

        private void WriteIntValue(int? value)
        {
            WriteValue(intValidator, value);
        }

        private readonly Validator intValidator = Validator_int.Get(0);

        private void WriteNoneValue()
        {
            WriteValue(noneValidator, NONE);
        }

        private readonly Validator noneValidator = Validator_none.Get();

        #endregion

        ///////////////////////////
        // LOCAL UTILITY METHODS //
        ///////////////////////////

        #region Local Utility Methods

        /// <summary>
        /// 
        /// </summary>
        /// <param name="v">the validator for this value, or none if there
        /// isn't one</param>
        /// <param name="value"></param>
        /// Exception:
        ///     throws Exception
        public void WriteValue(Validator v, Object value)
        {
            sbyte type = CheckValue(value);



            buf.PutByte(type);

            switch (type)
            {
                case TypeCode.NULL:
                case TypeCode.BOOLEAN_FALSE:
                case TypeCode.BOOLEAN_TRUE:
                case TypeCode.EMPTY_STRING:
                case TypeCode.NONE:
                    return;

                case TypeCode.BYTE:
                    buf.PutByte((Convert.ToSByte(value)));
                    return;
                case TypeCode.SHORT:
                    buf.PutShort(Convert.ToInt16(value));
                    return;

                case TypeCode.INT:
                    buf.PutInt(Convert.ToInt32(value));
                    return;

                case TypeCode.LONG:
                    buf.PutLong(Convert.ToInt64(value));
                    return;

                case TypeCode.FLOAT:
                    buf.PutFloat(Convert.ToSingle(value));
                    return;

                case TypeCode.DOUBLE:
                    buf.PutDouble(Convert.ToDouble(value));
                    return;

                case TypeCode.BYTES:
                    WriteBytes((sbyte[])value);
                    return;

                // reserved for future use
                //case TypeCode.BOOLS:
                //case TypeCode.SHORTS:
                //case TypeCode.INTS:
                //case TypeCode.LONGS:
                //case TypeCode.FLOATS:
                //case TypeCode.DOUBLES:
                case TypeCode.ARRAY:
                    WriteArray(ToArrayValue(value, v), v);
                    return;

                case TypeCode.STRING:
                    WriteBytes(Encoding.UTF8.GetBytes((String)value));
                    //            WriteBytes(Encoding.UTF8.GetBytes(Convert.ToString(value)));
                    return;


                case TypeCode.CUSTOM:
                    {
                        StructValue sv = vf.ExportCustomValue(value);
                        if (sv == null)
                            throw new Exception("unsupported type " + value.GetType());

                        WriteStruct(sv);
                        return;
                    }

                default:
                    // type is either "tiny" integer or unused...
                    if (type >= TypeCode.MIN_TINY_INT && type <= TypeCode.MAX_TINY_INT)
                        return;

                    throw new Exception("unsupported typecode " + type);
            }

        }

        /// <summary>
        /// Writes a value (signed or unsigned) to the output stream.
        /// </summary>
        /// <param name="value">the value to write to the output stream.</param>
        /// Exception:
        ///     throws Exception if there's a problem with the output stream
        private void WriteBytes(sbyte[] value)
        {
            int n = value.Length;
            WriteIntValue(n);
            //     WriteValue( intValidator, n );

      /*     byte[] bvalue = new byte[Buffer.ByteLength(value)];
            Buffer.BlockCopy(value, 0, bvalue, 0, Buffer.ByteLength(value)); */

            byte[] bvalue = (byte[])(Array)value;
           

            buf.Put(bvalue, 0, n);
        }

        /// <summary>
        /// Writes a value (signed or unsigned) to the output stream.
        /// </summary>
        /// <param name="value">the value to write to the output stream.</param>
        /// Exception:
        ///     throws Exception if there's a problem with the output stream
        private void WriteBytes(byte[] value)
        {
            int n = value.Length;
            WriteValue(intValidator, n);
            buf.Put(value, 0, n);
        }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="msg"></param>
        /// <param name="vf"></param>
        /// <returns></returns>
        /// Exception:
        ///     throws Exception
        public static byte[] GetBytes(Message msg, ValueFactory vf)
        {
            FlexBuffer buf = new FlexBuffer();
            BinaryTaggedDataOutput tdo = new BinaryTaggedDataOutput(vf);
            tdo.WriteMessage(msg, buf);
            return buf.GetAvailBytes();
        }

        #endregion Local utility methods
    }
}

File: src/etch/bindings/java/transport/fmt/binary/BinaryTaggedDataOutput.java
/* $Id$
 *
 * Created by sccomer on Jan 17, 2007.
 *
 * Copyright (c) 2007 Cisco Systems, Inc. All rights reserved.
 */

package etch.bindings.java.transport.fmt.binary;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import metreos.util.FlexBuffer;
import metreos.util.URL;
import etch.bindings.java.msg.ArrayValue;
import etch.bindings.java.msg.Field;
import etch.bindings.java.msg.Message;
import etch.bindings.java.msg.StructValue;
import etch.bindings.java.msg.TaggedDataOutput;
import etch.bindings.java.msg.Type;
import etch.bindings.java.msg.Validator;
import etch.bindings.java.msg.Validator_int;
import etch.bindings.java.msg.ValueFactory;
import etch.bindings.java.transport.fmt.TypeCode;

/**
 * Description of TaggedDataOutputStream.
 */
final public class BinaryTaggedDataOutput extends BinaryTaggedData
	implements TaggedDataOutput
{
	/**
	 * Constructs the BinaryTaggedDataOutput.
	 *
	 * @param buf the output buffer.
	 * @param vf the value factory for the service. 
	 */
	public BinaryTaggedDataOutput( FlexBuffer buf, ValueFactory vf )
	{
		super( vf );
		this.buf = buf;
	}
	
	/**
	 * Constructs the BinaryTaggedDataOutput with a null buffer.
	 * 
	 * @param vf the value factory for the service.
	 * @param uri the uri used to construct the transport stack.
	 * 
	 * @see #setBuffer(FlexBuffer)
	 */
	public BinaryTaggedDataOutput( ValueFactory vf, URL uri )
	{
		super( vf );
		// don't have anything to do with uri yet.
	}

	private FlexBuffer buf;

	//////////////////////////////
	// TaggedDataOutput methods //
	//////////////////////////////

	public void startMessage( Message msg ) throws IOException
	{
		// since message is top-level, no type code is written to
		// indicate a message is starting. we do write a version
		// number to indicate this version of the binary tagged
		// data output.
		structs.clear();
		buf.putByte( VERSION );
		writeStructType( msg.type() );
		pushStruct( msg );
	}

	public void endMessage( Message msg ) throws IOException
	{
		writeValue( noneValidator, NONE );
		popStruct( msg );
		assert structs.isEmpty();
	}

	public void startStruct( StructValue struct ) throws IOException
	{
		// the caller has already written a type code to indicate a
		// struct is starting.
		writeStructType( struct.type() );
		pushStruct( struct );
	}

	public void writeStructElement( Field key, Object value ) throws IOException
	{
		Validator v = currentType.getValidator( key );
		writeValue( intValidator, key.getId() );
		writeValue( v, value );
	}

	public void endStruct( StructValue struct ) throws IOException
	{
		writeValue( noneValidator, NONE );
		popStruct( struct );
	}
	
	private void pushStruct( StructValue struct )
	{
		structs.add( currentStruct );
		currentStruct = struct;
		currentType = currentStruct.type();
	}
	
	private void popStruct( StructValue struct )
	{
		assert currentStruct == struct;
		currentStruct = structs.remove( structs.size()-1 );
		currentType = currentStruct != null ? currentStruct.type() : null;
	}
	
	private final List<StructValue> structs = new ArrayList<StructValue>();
	
	private StructValue currentStruct;
	
	private Type currentType;

	public void startArray( ArrayValue array ) throws IOException
	{
		// the caller has already written a type code to indicate an
		// array is starting.
		byte type = array.typeCode();
		buf.putByte( type );
		if (type == TypeCode.CUSTOM)
			writeStructType( array.customStructType() );
		writeValue( intValidator, array.dim() );
	}
	
	private void writeStructType( Type type ) throws IOException
	{
		writeValue( intValidator, type.getId() );
	}

	public void writeArrayElement( Object value ) throws IOException
	{
		writeValue( arrayElementValidator, value );
	}
	
	private Validator arrayElementValidator;

	public void endArray( ArrayValue array ) throws IOException
	{
		writeValue( noneValidator, NONE );
	}
	
	private final Validator intValidator = Validator_int.get( 0 );
	
	private final Validator noneValidator = Validator_none.get();

	///////////////////////////
	// LOCAL UTILITY METHODS //
	///////////////////////////

	/**
	 * @param v the validator for this value, or none if there
	 * or none if there isn't one.
	 * @param value
	 * @throws IOException
	 */
	public void writeValue( Validator v, Object value ) throws IOException
	{
//		System.out.printf( "writeValue %s: %s\n", value, v );
		
		byte type = checkValue( value, v );
		
		if (type == TypeCode.ANY)
			type = checkValue( value );
		
		// xTODO disable this, we don't need it except as an
		// extra safety net during development.
		if (false)
		{
			if (v == null)
				throw new IOException( String.format(
					"value %s (class %s) has no validator",
					value, value.getClass() ) );
			
			if (value != null && !v.validate( value ))
				throw new IllegalArgumentException( String.format(
						"value %s (class %s) not valid for validator %s",
						value, value.getClass(), v ) );
			
			byte oldType = checkValue( value );
			if (type != oldType)
				throw new IOException( String.format(
					"value %s (class %s) type code %d, old type code %d",
					value, value != null ? value.getClass() : null, type, oldType ) );
		}
		
		buf.putByte( type );

		switch (type)
		{
			case TypeCode.NULL:
			case TypeCode.BOOLEAN_FALSE:
			case TypeCode.BOOLEAN_TRUE:
			case TypeCode.EMPTY_STRING:
			case TypeCode.NONE:
				return;
			
			case TypeCode.BYTE1:
			case TypeCode.SHORT1:
			case TypeCode.INT1:
			case TypeCode.LONG1:
				buf.putByte( ((Number) value).byteValue() );
				return;
			
			case TypeCode.SHORT2:
			case TypeCode.INT2:
			case TypeCode.LONG2:
				buf.putShort( ((Number) value).shortValue() );
				return;
			
			case TypeCode.INT4:
			case TypeCode.LONG4:
				buf.putInt( ((Number) value).intValue() );
				return;
			
			case TypeCode.LONG8:
				buf.putLong( ((Number) value).longValue() );
				return;
			
			case TypeCode.FLOAT4:
				buf.putFloat( ((Number) value).floatValue() );
				return;
			
			case TypeCode.FLOAT8:
				buf.putDouble( ((Number) value).doubleValue() );
				return;
			
			case TypeCode.BYTES:
				writeBytes( (byte[]) value );
				return;
			
			case TypeCode.STRING:
				writeBytes( ((String) value).getBytes( vf.getStringEncoding() ) );
				return;
			
			case TypeCode.STRUCT:
				((StructValue) value).writeStruct( this );
				return;
			
			case TypeCode.ARRAY:
			{
				Validator oldArrayElementValidator = arrayElementValidator;
				arrayElementValidator = v != null ? v.elementValidator() : null;
				
				if (value instanceof ArrayValue)
					((ArrayValue) value).write( this );
				else
					toArrayValue( value ).write( this );
				
				arrayElementValidator = oldArrayElementValidator;
				return;
			}
				
			case TypeCode.CUSTOM:
			{
				StructValue struct = vf.exportCustomValue( value );
				
				if (struct == null)
					throw new UnsupportedOperationException( "unsupported type "+value.getClass() );
				
				struct.writeStruct( this );
				return;
			}
			
			default:
				// type is either "small" integer or unused...
				return;
		}
	}

	/**
	 * Writes a value (signed or unsigned) to the output stream.
	 * @param value the value to write to the output stream.
	 * @throws IOException if there is a problem with the output stream.
	 */
	private void writeBytes( byte[] value ) throws IOException
	{
		int n = value.length;
		writeValue( intValidator, n );
		buf.put( value, 0, n );
	}

	/**
	 * @param msg
	 * @param vf the value factory for the service. 
	 * @return the bytes of the message
	 * @throws IOException
	 */
	public static byte[] getBytes( Message msg, ValueFactory vf ) throws IOException
	{
		FlexBuffer buf = new FlexBuffer();
		BinaryTaggedDataOutput tdo = new BinaryTaggedDataOutput( buf, vf );
		msg.writeMessage( tdo );
		return buf.getAvailBytes();
	}

	/**
	 * @param buf
	 */
	public void setBuffer( FlexBuffer buf )
	{
		this.buf = buf;
	}
}
